cmd/app/grpc.go

```diff
@@ -67,16 +70,85 @@ func PassFulcioConfigThruContext(cfg *config.FulcioConfig) grpc.UnaryServerInter
 	}
 }
 
-func createGRPCCreds(certPath, keyPath string) (grpc.ServerOption, error) {
-	cert, err := tls.LoadX509KeyPair(certPath, keyPath)
+type cachedTLSCert struct {
+	sync.RWMutex
+	certPath string
+	keyPath  string
+	cert     *tls.Certificate
+	Watcher  *fsnotify.Watcher
+}
+
+func newCachedTLSCert(certPath, keyPath string) (*cachedTLSCert, error) {
+	cachedTLSCert := &cachedTLSCert{
+		certPath: certPath,
+		keyPath:  keyPath,
+	}
+	if err := cachedTLSCert.UpdateCertificate(); err != nil {
+		return nil, err
+	}
+	var err error
+	cachedTLSCert.Watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-cachedTLSCert.Watcher.Events:
+				if !ok {
+					return
+				}
+				if event.Has(fsnotify.Write) {
+					log.Logger.Info("fsnotify grpc-tls-certificate write event detected")
+					if err := cachedTLSCert.UpdateCertificate(); err != nil {
+						log.Logger.Error(err)
+					}
+				}
+			case err, ok := <-cachedTLSCert.Watcher.Errors:
+				if !ok {
+					return
+				}
+				log.Logger.Error("fsnotify grpc-tls-certificate error:", err)
+			}
+		}
+	}()
+
+	// Add a path.
+	if err = cachedTLSCert.Watcher.Add(certPath); err != nil {
+		return nil, err
+	}
+	return cachedTLSCert, nil
+}
+
+func (c *cachedTLSCert) GetCertificate() *tls.Certificate {
+	// get reader lock
+	c.RLock()
```

Comment:
Do you expect this would have negligible performance impact? I would assume it's fine given it's a reader lock, but it is introducing a lock for every call.

Reply:
Yes, negligible. Getting a reader lock is simply an atomic check and
increment an int, and the writer lock should only block readers temporarily
for the file I/O to read the new cert from disk and parse it. Even if it’s
rotated every 24h that’s extremely minimal.